File: 3_Лаба/Исходники/Triangle.cpp
#include "Triangle.h"

#include "Line.h"
#include "GeometryMath.h"

Triangle::Triangle(const Point* points)
	: Shape(points, 3, "Triangle")
{}

double Triangle::FindArea() const
{
	GeometryMath math;
	
	Line* lines = FindEdges();

	double A = math.FindMagnitude(lines[0]);
	double B = math.FindMagnitude(lines[1]);
	double C = math.FindMagnitude(lines[2]);

	double p = (A + B + C) / 2;

	delete[] lines;
	return sqrt(p * (p - A) * (p - B) * (p - C));
}

bool Triangle::IsExisting() const
{
	GeometryMath math;

	Line* lines = FindEdges();

	double A = math.FindMagnitude(lines[0]);
	double B = math.FindMagnitude(lines[1]);
	double C = math.FindMagnitude(lines[2]);

	delete[] lines;

	if (A + B > C && A + C > B && B + C > A)
	{
		return true;
	}
	return false;
}

File: 5_6_Лаба/FiguresDrawer/Model/Figures/Triangle.cs
using FiguresDrawer.Model.Structures;

namespace FiguresDrawer.Model.Figures
{
	public class Triangle : FigureBase
	{
		public Triangle(Point[] points) : base(points)
		{
		}

		public override double FindArea()
		{
			var lines = GetLines();

			var p = FindPerimeter() / 2;
			var a = GeometryMath.FindMagnitude(lines[0]);
			var b = GeometryMath.FindMagnitude(lines[1]);
			var c = GeometryMath.FindMagnitude(lines[2]);

			return System.Math.Sqrt(p * (p - a) * (p - b) * (p - c));
		}

		public override bool IsExisting()
		{
			var lines = GetLines();

			var a = GeometryMath.FindMagnitude(lines[0]);
			var b = GeometryMath.FindMagnitude(lines[1]);
			var c = GeometryMath.FindMagnitude(lines[2]);

			return a + b > c && a + c > b && b + c > a;
		}

		private LineSegment[] GetLines()
		{
			var lines = new LineSegment[3];
			var points = GetPoints();

			for (int i = 0; i < lines.Length; i++)
			{
				lines[i] = new LineSegment(points[i], points[(i + 1) % lines.Length]);
			}

			return lines;
		}
	}
}
